proposed/2021/MachineReadableOutput.md

```diff
@@ -0,0 +1,637 @@
+
+* We won't support saving the machine-readable output to disk as part of this spec. The work-around is for the consumer to read from the console's stdout stream.
+* At this point, no other CLI commands (e.g. dotnet list reference) will be within scope for this feature.
+"--parsable" option needs separate spec.
+* Currently license info is not emitted from any cli command, it could be quite useful, we should consider in the future.
+
+## Rationale and alternatives
+Currently, no other `dotnet command` implemented this, this is the 1st time dotnet command implementing `json`(etc) output, so it could become example for others next time they implement.
+Please note, except "tab completion" (for dotnet) part all changes would be inside NuGet.Client repo(under NuGet.Core), and risk of introducing regression is low.(`--format text` refactoring related changes only come into my mind.), no impact on dotnet sdk.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevent to this proposal? -->
+
+* https://github.com/NuGet/Home/blob/dotnet-audit/proposed/2021/DotNetAudit.md#dotnet-audit---json
+
+* https://github.com/NuGet/Home/wiki/%5BSpec%5D-Machine-readable-output-for-dotnet-list-package
+
+## Unresolved Questions
+
+* Chris: One problem is most `dotnet list package` options outputs are exclusive and have to query separately and join result to get full picture. Both below approach require additional work.
+  * dotnet cli needs an all up `dotnet list package --all` that shows all deprecated, vulnerable, and outdated top level and transitive packages. [r766860629](https://github.com/NuGet/Home/pull/11446#discussion_r766860629)
+  * Alternatively the behavior of --outdated and --deprecated could be additive rather than exclusive. That adds to the scope of this work though.[r766860629](https://github.com/NuGet/Home/pull/11446#discussion_r766860629). With this approach we can have single schema to populate.
+
+* Donnie: When I want to create archival records, will I want something more unique than the project name?
+Adding the path, repo, commit ID, etc seems complex. [r766920783](https://github.com/NuGet/Home/pull/11446#discussion_r766920783)
+  * `name/relative path to solution` could be solution here.
+
+* Donnie: How can we record in the output that --include-transitive wasn't used here?
+In other words, if I look at this output years from now, how would I know whether any transitives were in this project? [r766924390](https://github.com/NuGet/Home/pull/11446#discussion_r766924390)
+  * packages.lock.json format could be used here.  
+
+* Loïc : Should `dotnet list package` include hashes or package source for each dependency? The package ID and version isn't globally unique across package sources?
```

Comment:
> package source

This is probably a good idea. The issue with hashes is a really good point re:re-signing. Does that happen often? If it does are hashes published for previous versions of re-signed packages in vuln reports or similar from the msft side?

> The package ID and version isn't globally unique across package sources

In this case is there really anything you can do? I mean technically there isn't even strong consistency within a single package source if we're already in the realm of mismanaging them. If a local nuget server admin nukes a package and then allows republishing of a second one I think at that it's a bit of 'you're on your own' territory and that's on the consumer of that package. 

AFAIK that's not a workflow nuget really supports (I could be wrong). I feel like if you're publishing the same version of a package with different contents in two places you're implicitly accepting that you're outside the safe zone. That being said I think that the package source + the file hash wouldn't be a bad idea to include.

I think that the mismanagement of nuget servers isn't something that the nuget team can/should care about but providing more information in the output is probably better. The re-signing one is the only one that really gives me the heebies because it could potentially give you a false sense of security re:vulns etc but it might be something that has to be solved by external tooling.